File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/StudioContext.java
package fr.hardel.asset_editor.client.javafx.lib;

import fr.hardel.asset_editor.client.ClientSessionDispatch;
import fr.hardel.asset_editor.client.javafx.lib.action.EditorActionGateway;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioConcept;
import fr.hardel.asset_editor.store.EnchantmentFlushAdapter;
import fr.hardel.asset_editor.client.javafx.lib.data.StudioViewMode;
import fr.hardel.asset_editor.client.javafx.routes.StudioRoute;
import fr.hardel.asset_editor.client.javafx.routes.StudioRouter;
import fr.hardel.asset_editor.client.selector.SelectorEquality;
import fr.hardel.asset_editor.client.selector.StoreSelection;
import fr.hardel.asset_editor.client.selector.Subscription;
import fr.hardel.asset_editor.client.state.ClientSessionState;
import fr.hardel.asset_editor.client.state.ClientWorkspaceState;
import fr.hardel.asset_editor.client.state.ClientPackInfo;
import fr.hardel.asset_editor.client.state.RegistryElementStore;
import fr.hardel.asset_editor.client.state.StudioOpenTab;
import fr.hardel.asset_editor.client.state.WorkspacePackSelectionState;
import fr.hardel.asset_editor.client.state.WorkspaceTabsState;
import fr.hardel.asset_editor.client.state.WorkspaceUiState;
import fr.hardel.asset_editor.store.ElementEntry;
import fr.hardel.asset_editor.permission.StudioPermissions;
import net.minecraft.client.Minecraft;
import net.minecraft.core.Holder;
import net.minecraft.core.HolderSet;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.Registries;
import net.minecraft.resources.Identifier;
import net.minecraft.resources.ResourceKey;
import net.minecraft.tags.TagKey;

import java.util.List;
import java.util.Optional;

public final class StudioContext {

    private final ClientSessionState sessionState;
    private final ClientWorkspaceState workspaceState;
    private final ClientSessionDispatch dispatch;
    private final StudioRouter router = new StudioRouter();
    private final EditorActionGateway gateway;
    private final Subscription permissionSubscription;
    private final Subscription packSelectionSubscription;

    public StudioContext(ClientSessionState sessionState, ClientSessionDispatch dispatch) {
        this.sessionState = sessionState;
        this.workspaceState = new ClientWorkspaceState(sessionState);
        this.dispatch = dispatch;
        this.gateway = new EditorActionGateway(sessionState, workspaceState);
        this.permissionSubscription = sessionState.select(ClientSessionState.Snapshot::permissions)
            .subscribe(permission -> enforcePermissionRoute(), true);
        this.packSelectionSubscription = workspaceState.packSelectionState()
            .select(WorkspacePackSelectionState.Snapshot::selectedPack)
            .subscribe(pack -> refreshSelectedPackWorkspace(), true);

        router.setPermissionSupplier(sessionState::permissions);
        dispatch.setGateway(gateway);
    }

    public ClientSessionState sessionState() {
        return sessionState;
    }

    public ClientWorkspaceState workspaceState() {
        return workspaceState;
    }

    public StudioRouter router() {
        return router;
    }

    public WorkspaceUiState uiState() {
        return workspaceState.uiState();
    }

    public WorkspaceTabsState tabsState() {
        return workspaceState.tabsState();
    }

    public WorkspacePackSelectionState packState() {
        return workspaceState.packSelectionState();
    }

    public RegistryElementStore elementStore() {
        return workspaceState.elementStore();
    }

    public EditorActionGateway gateway() {
        return gateway;
    }

    public StudioPermissions permissions() {
        return sessionState.permissions();
    }

    public StoreSelection<ClientSessionState.Snapshot, StudioPermissions> selectPermissions() {
        return sessionState.select(ClientSessionState.Snapshot::permissions);
    }

    public StoreSelection<ClientSessionState.Snapshot, List<ClientPackInfo>> selectAvailablePacks() {
        return sessionState.select(ClientSessionState.Snapshot::availablePacks, SelectorEquality.listEquality());
    }

    public StoreSelection<WorkspacePackSelectionState.Snapshot, ClientPackInfo> selectSelectedPack() {
        return workspaceState.packSelectionState().select(WorkspacePackSelectionState.Snapshot::selectedPack);
    }

    public StoreSelection<WorkspaceTabsState.Snapshot, List<StudioOpenTab>> selectOpenTabs() {
        return workspaceState.tabsState().select(WorkspaceTabsState.Snapshot::openTabs, SelectorEquality.listEquality());
    }

    public StoreSelection<WorkspaceTabsState.Snapshot, Integer> selectActiveTabIndex() {
        return workspaceState.tabsState().select(WorkspaceTabsState.Snapshot::activeTabIndex);
    }

    public StoreSelection<WorkspaceTabsState.Snapshot, String> selectCurrentElementId() {
        return workspaceState.tabsState().select(WorkspaceTabsState.Snapshot::currentElementId);
    }

    public StoreSelection<WorkspaceUiState.Snapshot, String> selectFilterPath() {
        return workspaceState.uiState().select(WorkspaceUiState.Snapshot::filterPath);
    }

    public StoreSelection<WorkspaceUiState.Snapshot, StudioViewMode> selectViewMode() {
        return workspaceState.uiState().select(WorkspaceUiState.Snapshot::viewMode);
    }

    private void enforcePermissionRoute() {
        if (router.currentRoute() == StudioRoute.NO_PERMISSION) {
            StudioConcept.firstAccessible(permissions()).ifPresent(concept -> router.navigate(concept.overviewRoute()));
            return;
        }
        router.revalidate();
    }

    public <T> List<ElementEntry<T>> allTypedEntries(ResourceKey<Registry<T>> registryKey) {
        return workspaceState.elementStore().allTypedElements(registryKey);
    }

    public <T> ElementEntry<T> currentEntry(ResourceKey<Registry<T>> registryKey) {
        String id = workspaceState.tabsState().currentElementId();
        if (id == null || id.isBlank())
            return null;
        Identifier identifier = Identifier.tryParse(id);
        if (identifier == null)
            return null;
        return workspaceState.elementStore().get(registryKey, identifier);
    }

    public <T> List<Holder.Reference<T>> registryElements(ResourceKey<Registry<T>> registryKey) {
        var conn = Minecraft.getInstance().getConnection();
        if (conn == null)
            return List.of();
        return conn.registryAccess()
            .lookup(registryKey)
            .map(registry -> registry.listElements().toList())
            .orElse(List.of());
    }

    public <T> Optional<HolderSet<T>> resolveTag(ResourceKey<Registry<T>> registryKey, TagKey<T> tagKey) {
        var conn = Minecraft.getInstance().getConnection();
        if (conn == null)
            return Optional.empty();
        return conn.registryAccess().lookup(registryKey)
            .flatMap(registry -> registry.get(tagKey))
            .map(named -> (HolderSet<T>) named);
    }

    public <T> Optional<Holder.Reference<T>> resolveHolder(ResourceKey<Registry<T>> registryKey, Identifier id) {
        var conn = Minecraft.getInstance().getConnection();
        if (conn == null)
            return Optional.empty();
        return conn.registryAccess().lookup(registryKey).flatMap(registry -> registry.get(ResourceKey.create(registryKey, id)));
    }

    public void resyncWorldSession(boolean force) {
        String nextKey = sessionState.worldSessionKey();
        if (workspaceState.worldSessionKey().equals(nextKey)) {
            if (force)
                sessionState.refreshPackList();
            return;
        }

        workspaceState.setWorldSessionKey(nextKey);
        workspaceState.resetForWorldSync();
        sessionState.refreshPackList();
        snapshotRegistries();
    }

    public void resetForWorldClose() {
        workspaceState.resetForWorldClose();
        dispatch.clearGateway(gateway);
        router.navigate(StudioRoute.NO_PERMISSION);
    }

    private void snapshotRegistries() {
        var conn = Minecraft.getInstance().getConnection();
        if (conn == null)
            return;
        conn.registryAccess().lookup(Registries.ENCHANTMENT).ifPresent(registry ->
            workspaceState.elementStore().snapshotFromRegistry(
                Registries.ENCHANTMENT,
                registry,
                entry -> EnchantmentFlushAdapter.initializeCustom(entry.data(), entry.tags())));
    }

    public void dispose() {
        permissionSubscription.unsubscribe();
        packSelectionSubscription.unsubscribe();
        dispatch.clearGateway(gateway);
        workspaceState.dispose();
    }

    private void refreshSelectedPackWorkspace() {
        workspaceState.clearPendingActions();
        workspaceState.issueState().clear();
        if (workspaceState.packSelectionState().selectedPack() == null)
            return;
        gateway.requestPackWorkspace(Registries.ENCHANTMENT);
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/StudioContext.kt
package fr.hardel.asset_editor.client.compose.lib

import fr.hardel.asset_editor.client.ClientSessionDispatch
import fr.hardel.asset_editor.client.compose.lib.action.EditorActionGateway
import fr.hardel.asset_editor.client.compose.lib.assets.DefaultStudioAssetCache
import fr.hardel.asset_editor.client.compose.lib.assets.DefaultStudioPrefetcher
import fr.hardel.asset_editor.client.compose.lib.assets.StudioAssetCache
import fr.hardel.asset_editor.client.compose.lib.assets.StudioPrefetcher
import fr.hardel.asset_editor.client.debug.ClientDebugTelemetry
import fr.hardel.asset_editor.client.memory.core.Subscription
import fr.hardel.asset_editor.client.memory.session.debug.DebugMemory
import fr.hardel.asset_editor.client.memory.session.ui.NavigationMemory

import fr.hardel.asset_editor.client.memory.session.SessionMemory
import fr.hardel.asset_editor.client.memory.core.ClientWorkspaceRegistries
import fr.hardel.asset_editor.client.memory.core.ClientWorkspaceRegistry
import fr.hardel.asset_editor.client.memory.session.ui.UiMemory
import fr.hardel.asset_editor.client.ClientPreferences
import fr.hardel.asset_editor.client.memory.persistent.IssueMemory
import fr.hardel.asset_editor.client.memory.session.ui.PackSelectionMemory
import fr.hardel.asset_editor.client.memory.session.server.RegistryMemory
import fr.hardel.asset_editor.permission.StudioPermissions
import fr.hardel.asset_editor.data.concept.StudioRegistryResolver
import java.util.Optional
import net.minecraft.client.Minecraft
import net.minecraft.client.resources.language.I18n
import net.minecraft.core.Holder
import net.minecraft.core.HolderSet
import net.minecraft.core.Registry
import net.minecraft.resources.Identifier
import net.minecraft.resources.ResourceKey
import net.minecraft.tags.TagKey

class StudioContext(
    private val sessionMemory: SessionMemory,
    private val debugMemory: DebugMemory,
    private val dispatch: ClientSessionDispatch
) {

    private val packSelection =
        PackSelectionMemory(sessionMemory, ClientPreferences::lastPackId, ClientPreferences::setLastPackId)
    private val issues = IssueMemory()
    private val registries = RegistryMemory()
    private val subscriptions = mutableListOf<Subscription>()
    private val navigationMemory = NavigationMemory(sessionMemory::permissions)
    private val uiMemory = UiMemory()
    private val assetCache = DefaultStudioAssetCache()
    private val prefetcher = DefaultStudioPrefetcher(assetCache)
    private var lastWorldSessionKey = sessionMemory.worldSessionKey()

    val gateway = EditorActionGateway(sessionMemory, packSelection, registries, issues)

    init {
        snapshotRegistries()

        var lastPermissions = sessionMemory.permissions()
        fun handlePermissions(permissions: StudioPermissions) {
            navigationMemory.revalidate(permissions)
            if (navigationMemory.snapshot().current is NoPermissionDestination) {
                if (!permissions.isNone) {
                    StudioUiRegistry.firstSupportedConceptId()?.let { conceptId ->
                        navigationMemory.navigate(ConceptOverviewDestination(conceptId))
                    }
                }
            }
        }
        subscriptions += sessionMemory.subscribe {
            val permissions = sessionMemory.permissions()
            if (permissions == lastPermissions) {
                return@subscribe
            }
            lastPermissions = permissions
            handlePermissions(permissions)
        }
        handlePermissions(lastPermissions)

        var lastSelectedPackId = packSelection.selectedPack()?.packId()
        subscriptions += packSelection.subscribe {
            val selectedPackId = packSelection.selectedPack()?.packId()
            if (selectedPackId == lastSelectedPackId) {
                return@subscribe
            }
            lastSelectedPackId = selectedPackId
            refreshSelectedPackWorkspace()
        }
        refreshSelectedPackWorkspace()

        dispatch.setGateway(gateway)
    }

    fun sessionMemory(): SessionMemory = sessionMemory

    fun packSelectionMemory(): PackSelectionMemory = packSelection

    fun issueMemory(): IssueMemory = issues

    fun registryMemory(): RegistryMemory = registries

    fun navigationMemory(): NavigationMemory = navigationMemory

    fun uiMemory(): UiMemory = uiMemory

    fun debugMemory(): DebugMemory = debugMemory

    fun assetCache(): StudioAssetCache = assetCache

    fun prefetcher(): StudioPrefetcher = prefetcher

    fun <T : Any> resolveTag(registryKey: ResourceKey<Registry<T>>, tagKey: TagKey<T>): Optional<HolderSet<T>> {
        val connection = Minecraft.getInstance().connection ?: return Optional.empty()
        return connection.registryAccess()
            .lookup(registryKey)
            .flatMap { registry -> registry.get(tagKey) }
            .map { named -> named as HolderSet<T> }
    }

    fun <T : Any> resolveHolder(registryKey: ResourceKey<Registry<T>>, id: Identifier): Optional<Holder.Reference<T>> {
        val connection = Minecraft.getInstance().connection ?: return Optional.empty()
        return connection.registryAccess()
            .lookup(registryKey)
            .flatMap { registry -> registry.get(ResourceKey.create(registryKey, id)) }
    }

    fun registryAccess() = Minecraft.getInstance().connection?.registryAccess()

    fun requireRegistryAccess() = registryAccess() ?: error("Missing client registry access")

    fun studioConceptId(registryKey: ResourceKey<out Registry<*>>) =
        registryAccess()?.let { StudioRegistryResolver.conceptId(it, registryKey) }

    fun studioDefaultEditorTab(conceptId: Identifier): Identifier =
        StudioRegistryResolver.defaultEditorTab(requireRegistryAccess(), conceptId)

    fun studioEditorTabs(conceptId: Identifier): List<Identifier> =
        StudioRegistryResolver.editorTabs(requireRegistryAccess(), conceptId)

    fun studioRegistryKey(conceptId: Identifier): ResourceKey<out Registry<*>> =
        StudioRegistryResolver.registryKey(requireRegistryAccess(), conceptId)

    fun studioRegistryPath(conceptId: Identifier): String =
        StudioRegistryResolver.registryPath(requireRegistryAccess(), conceptId)

    fun studioTitleKey(conceptId: Identifier): String =
        StudioRegistryResolver.titleKey(requireRegistryAccess(), conceptId)

    fun studioTabTitleKey(conceptId: Identifier, tabId: Identifier): String =
        StudioRegistryResolver.tabTitleKey(requireRegistryAccess(), conceptId, tabId)

    fun studioIcon(conceptId: Identifier): Identifier =
        StudioRegistryResolver.icon(requireRegistryAccess(), conceptId)

    fun resyncWorldSession() {
        val nextKey = sessionMemory.worldSessionKey()
        if (lastWorldSessionKey == nextKey) {
            return
        }

        ClientDebugTelemetry.lifecycle(
            I18n.get("debug:telemetry.world_session_changed"),
            mapOf(
                "previousWorldSessionKey" to lastWorldSessionKey,
                "nextWorldSessionKey" to nextKey
            )
        )

        lastWorldSessionKey = nextKey
        registries.clearAll()
        packSelection.clearSelection()
        issues.clear()
        gateway.clearPendingActions()
        snapshotRegistries()
    }

    fun resetForWorldClose() {
        lastWorldSessionKey = ""
        registries.clearAll()
        packSelection.clearSelection()
        issues.clear()
        gateway.clearPendingActions()
        uiMemory.reset()
        navigationMemory.reset()
        assetCache.invalidateAll()
        dispatch.clearGateway(gateway)
    }

    fun dispose() {
        subscriptions.forEach(Subscription::unsubscribe)
        assetCache.invalidateAll()
        dispatch.clearGateway(gateway)
        packSelection.dispose()
    }

    private fun snapshotRegistries() {
        val connection = Minecraft.getInstance().connection ?: return
        for (workspace in ClientWorkspaceRegistries.all()) {
            snapshotWorkspace(connection.registryAccess(), workspace)
        }
    }

    private fun refreshSelectedPackWorkspace() {
        gateway.clearPendingActions()
        issues.clear()
        val pack = packSelection.selectedPack() ?: return

        for (workspace in ClientWorkspaceRegistries.all()) {
            requestWorkspaceRefresh(pack.packId(), workspace)
        }
    }

    private fun requestWorkspaceRefresh(
        packId: String,
        workspace: ClientWorkspaceRegistry<*>
    ) {
        ClientDebugTelemetry.sync(
            I18n.get("debug:telemetry.requested_workspace_refresh"),
            mapOf(
                "packId" to packId,
                "registry" to workspace.registryId().toString()
            )
        )

        gateway.requestPackWorkspace(workspace)
    }

    private fun snapshotWorkspace(
        registryAccess: net.minecraft.core.RegistryAccess,
        workspace: ClientWorkspaceRegistry<*>
    ) {
        snapshotWorkspaceTyped(registryAccess, workspace)
    }

    private fun <T : Any> snapshotWorkspaceTyped(
        registryAccess: net.minecraft.core.RegistryAccess,
        workspace: ClientWorkspaceRegistry<T>
    ) {
        registryAccess.lookup(workspace.registryKey()).ifPresent { registry ->
            registries.snapshotFromRegistry(workspace, registry)
        }
    }
}
